## 1.Data Cleaning

### Download data dari 
https://github.com/realpython/python-data-cleaning/blob/master/Datasets/BL-Flickr-Images-Book.csv

In [15]:
## import

import pandas as pd
import numpy as np


In [16]:
df = pd.read_csv('BL-Flickr-Images-Book.csv')

In [17]:
df.describe()

Identifier  Corporate Author  Corporate Contributors  Engraver
count  8.287000e+03               0.0                     0.0       0.0
mean   2.017344e+06               NaN                     NaN       NaN
std    1.190379e+06               NaN                     NaN       NaN
min    2.060000e+02               NaN                     NaN       NaN
25%    9.157875e+05               NaN                     NaN       NaN
50%    2.043707e+06               NaN                     NaN       NaN
75%    3.047430e+06               NaN                     NaN       NaN
max    4.160339e+06               NaN                     NaN       NaN

In [18]:
df.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [19]:
df.dtypes

Identifier                  int64
Edition Statement          object
Place of Publication       object
Date of Publication        object
Publisher                  object
Title                      object
Author                     object
Contributors               object
Corporate Author          float64
Corporate Contributors    float64
Former owner               object
Engraver                  float64
Issuance type              object
Flickr URL                 object
Shelfmarks                 object
dtype: object

In [20]:
## Menghapus data NaN

yg_dihapus = ['Edition Statement','Corporate Author','Corporate Contributors','Former owner','Engraver','Shelfmarks']

df.drop(yg_dihapus, inplace=True, axis=1)

In [21]:
df.head()

Identifier      Place of Publication Date of Publication  \
0         206                    London         1879 [1878]   
1         216  London; Virtue & Yorston                1868   
2         218                    London                1869   
3         472                    London                1851   
4         480                    London                1857   

               Publisher                                              Title  \
0       S. Tinsley & Co.                  Walter Forbes. [A novel.] By A. A   
1           Virtue & Co.  All for Greed. [A novel. The dedication signed...   
2  Bradbury, Evans & Co.  Love the Avenger. By the author of “All for Gr...   
3          James Darling  Welsh Sketches, chiefly ecclesiastical, to the...   
4   Wertheim & Macintosh  [The World in which I live, and my place in it...   

      Author                                  Contributors Issuance type  \
0      A. A.                               FORBES, Walter.   monographic   
1  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
2  A., A. A.  BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
3  A., E. S.                   Appleyard, Ernest Silvanus.   monographic   
4  A., E. S.                           BROOME, John Henry.   monographic   

                                          Flickr URL  
0  http://www.flickr.com/photos/britishlibrary/ta...  
1  http://www.flickr.com/photos/britishlibrary/ta...  
2  http://www.flickr.com/photos/britishlibrary/ta...  
3  http://www.flickr.com/photos/britishlibrary/ta...  
4  http://www.flickr.com/photos/britishlibrary/ta...

## 2.Mengubah indeks di data


In [22]:
df['Identifier'].is_unique

True

In [23]:
df = df.set_index('Identifier')

In [24]:
df.head()

Place of Publication Date of Publication  \
Identifier                                                 
206                           London         1879 [1878]   
216         London; Virtue & Yorston                1868   
218                           London                1869   
472                           London                1851   
480                           London                1857   

                        Publisher  \
Identifier                          
206              S. Tinsley & Co.   
216                  Virtue & Co.   
218         Bradbury, Evans & Co.   
472                 James Darling   
480          Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                            Contributors Issuance type  \
Identifier                                                               
206                                      FORBES, Walter.   monographic   
216         BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
218         BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
472                          Appleyard, Ernest Silvanus.   monographic   
480                                  BROOME, John Henry.   monographic   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

In [28]:
df.loc[472]

Place of Publication                                               London
Date of Publication                                                  1851
Publisher                                                   James Darling
Title                   Welsh Sketches, chiefly ecclesiastical, to the...
Author                                                          A., E. S.
Contributors                                  Appleyard, Ernest Silvanus.
Issuance type                                                 monographic
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 472, dtype: object

## 3.Merapikan Fields

In [30]:
## menghitung jumlah objects
df.dtypes.value_counts()

object    8
dtype: int64

In [36]:
df.loc[1905:,'Date of Publication'].head(15)

Identifier
1905                  1888
1929           1839, 38-54
2836                  1897
2854                  1865
2956               1860-63
2957                  1873
3017                  1866
3131                  1899
4598                  1814
4884                  1820
4976                  1800
5382    1847, 48 [1846-48]
5385               [1897?]
5389               [1897?]
5432                  1893
Name: Date of Publication, dtype: object

~ hilangkan tanggal lain dalam kurung siku
~ konversi rentang tanggal
~ hilangkan tanggal gak jelas [1897?] --> NaN
~ konversi ke NaN


In [37]:
regex = r'^(\d{4})'

In [39]:
## ekstrak data ke 4 digit numerik
extraks = df['Date of Publication'].str.extract(r'^(\d{4})', expand=False)
extraks.head(15)

Identifier
206     1879
216     1868
218     1869
472     1851
480     1857
481     1875
519     1872
667      NaN
874     1676
1143    1679
1280    1802
1808    1859
1905    1888
1929    1839
2836    1897
Name: Date of Publication, dtype: object

In [40]:
## dtype masih object --> angka

df['Date of Publication'] = pd.to_numeric(extraks)
df['Date of Publication'].dtype

dtype('float64')

## 4.membersihkan kolom

In [41]:
## mencari kolom yang perlu dibuang

# 'Place of Publication'

df['Place of Publication'].head(15)

## ada lokasi tapi nama halaman dan beberapa tidak rapi

Identifier
206                                  London
216                London; Virtue & Yorston
218                                  London
472                                  London
480                                  London
481                                  London
519                                  London
667     pp. 40. G. Bryan & Co: Oxford, 1898
874                                 London]
1143                                 London
1280                               Coventry
1808                            Christiania
1905                                Firenze
1929                              Amsterdam
2836                                 Savona
Name: Place of Publication, dtype: object

In [42]:
df.loc[667]

Place of Publication                  pp. 40. G. Bryan & Co: Oxford, 1898
Date of Publication                                                   NaN
Publisher                                                             NaN
Title                   The Coming of Spring, and other poems. By J. A...
Author                                                      A., J.|A., J.
Contributors                                ANDREWS, J. - Writer of Verse
Issuance type                                                 monographic
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 667, dtype: object

In [46]:
df.loc[874]

Place of Publication                                              London]
Date of Publication                                                1676.0
Publisher                                                             NaN
Title                   A Warning to the inhabitants of England, and L...
Author                                                             Remaʿ.
Contributors                                                 ADAMS, Mary.
Issuance type                                                 monographic
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 874, dtype: object

In [47]:
df.loc[4115138]

Place of Publication                                 G. Eld: London, 1608
Date of Publication                                                   NaN
Publisher                                                             NaN
Title                   The Revenger's Tagaedie. As it hath beene sund...
Author                                                                NaN
Contributors                                             TOURNEUR, Cyril.
Issuance type                                                 monographic
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 4115138, dtype: object

In [48]:
## Merapikan dengan string contains

publikasi = df['Place of Publication']

london = publikasi.str.contains('London')
london[:5]


Identifier
206    True
216    True
218    True
472    True
480    True
Name: Place of Publication, dtype: bool

In [49]:
oxford = publikasi.str.contains('Oxford')

In [50]:
## membersihkan dengan np where
df['Place of Publication'] = np.where(london, 'London', np.where(oxford, 'Oxford', publikasi.str.replace('-',' ')))

In [53]:
df['Place of Publication'].tail(15)

Identifier
4115138                 London
4116063                 London
4117526                Leipzig
4117583                 London
4117749               New York
4117751               New York
4117752                 London
4156359               New York
4157746                 London
4157862    Newcastle upon Tyne
4158088                 London
4158128                  Derby
4159563                 London
4159587    Newcastle upon Tyne
4160339                 London
Name: Place of Publication, dtype: object

In [54]:
df.head()

Place of Publication  Date of Publication              Publisher  \
Identifier                                                                    
206                      London               1879.0       S. Tinsley & Co.   
216                      London               1868.0           Virtue & Co.   
218                      London               1869.0  Bradbury, Evans & Co.   
472                      London               1851.0          James Darling   
480                      London               1857.0   Wertheim & Macintosh   

                                                        Title     Author  \
Identifier                                                                 
206                         Walter Forbes. [A novel.] By A. A      A. A.   
216         All for Greed. [A novel. The dedication signed...  A., A. A.   
218         Love the Avenger. By the author of “All for Gr...  A., A. A.   
472         Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
480         [The World in which I live, and my place in it...  A., E. S.   

                                            Contributors Issuance type  \
Identifier                                                               
206                                      FORBES, Walter.   monographic   
216         BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
218         BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
472                          Appleyard, Ernest Silvanus.   monographic   
480                                  BROOME, John Henry.   monographic   

                                                   Flickr URL  
Identifier                                                     
206         http://www.flickr.com/photos/britishlibrary/ta...  
216         http://www.flickr.com/photos/britishlibrary/ta...  
218         http://www.flickr.com/photos/britishlibrary/ta...  
472         http://www.flickr.com/photos/britishlibrary/ta...  
480         http://www.flickr.com/photos/britishlibrary/ta...

## 5.Menggunakan Apply map dengan dataset baru
https://github.com/realpython/python-data-cleaning/blob/master/Datasets/university_towns.txt

In [56]:
university_town = []

with open("university_towns.txt") as file:
    for line in file:
        if '[edit]' in line:
            state = line
        else:
            university_town.append((state, line))

university_town[:5]

[('Alabama[edit]\n', 'Auburn (Auburn University)[1]\n'),
 ('Alabama[edit]\n', 'Florence (University of North Alabama)\n'),
 ('Alabama[edit]\n', 'Jacksonville (Jacksonville State University)[2]\n'),
 ('Alabama[edit]\n', 'Livingston (University of West Alabama)[2]\n'),
 ('Alabama[edit]\n', 'Montevallo (University of Montevallo)[2]\n')]

In [57]:
df_kota = pd.DataFrame(university_town, columns=['state','RegionName'])
df_kota.head()

state                                         RegionName
0  Alabama[edit]\n                    Auburn (Auburn University)[1]\n
1  Alabama[edit]\n           Florence (University of North Alabama)\n
2  Alabama[edit]\n  Jacksonville (Jacksonville State University)[2]\n
3  Alabama[edit]\n       Livingston (University of West Alabama)[2]\n
4  Alabama[edit]\n         Montevallo (University of Montevallo)[2]\n

In [60]:
## applymap

def get_citystate(item):
    if '(' in item:
        return item[:item.find('(')]
    
    elif '[' in item:
        return item[:item.find('[')]
    else:
        return item



In [61]:
df_kota = df_kota.applymap(get_citystate)
df_kota.head()

state    RegionName
0  Alabama        Auburn
1  Alabama      Florence
2  Alabama  Jacksonville
3  Alabama    Livingston
4  Alabama    Montevallo